apisix/plugins/traffic-split.lua

```diff
@@ -258,18 +258,21 @@ end
 local function new_rr_obj(weighted_upstreams)
     local server_list = {}
     for i, upstream_obj in ipairs(weighted_upstreams) do
-        if not upstream_obj.upstream then
+        if not upstream_obj.upstream and not upstream_obj.upstream_id then
             -- If the `upstream` object has only the `weight` value, it means
             -- that the `upstream` weight value on the default `route` has been reached.
-            -- Need to set an identifier to mark the empty upstream.
-            upstream_obj.upstream = "empty_upstream"
+            -- Mark empty upstream services in the plugin, but with weight.
+            upstream_obj.upstream = "plugin#upstream#is#empty"
+            server_list[upstream_obj.upstream] = upstream_obj.weight
         end
 
-        if type(upstream_obj.upstream) == "table" then
+        if upstream_obj.upstream_id then
+            server_list[upstream_obj.upstream_id] = upstream_obj.weight
+        elseif type(upstream_obj.upstream) == "table" then
```

Comment:
Seems the schema of upstream will guard the type of upstream and we don't check the type again here.

Reply:
updated.